### DISCUSSION TITLE
How to compose a Patrol finder to find a widget that contains two Text descendants?

### DISCUSSION THREAD

**Discussion by proximilian** [leancodepl/patrol#1133] [Q&A]
Hello guys,

I stuck on the following issue -> I would like to to tap on the first widget of a listView, but that widget must contain both e.g. 'text1' and 'text2'. These `Text` are both descendants of the widget. As far as I am familiar with Patrols documentation in case of only one `Text` I would use e.g.  `$(#ExampleListView).$(ExampleWidgetType).containing('text1').tap() ` 
My question is how to add the additional 'text2' that must be contained by the widget as well? Or any suggestion for a workaround?

Thanks in advance, guys :)

**Comment by bartekpacia** [comment]
Hi again, @proximilian :)

Does the following work for you?

```dart
$(#ExampleListView).$(ExampleWidgetType).containing('text1').containing('text2').tap();
```

**Reply by proximilian** [reply]
Thanks for the really quick response. 
I was thinking of it actually. Still haven't try it because I thought that in that way the second 'containing widget' will be considered as descendant of the previous one. Please feel free to correct me if I am wrong. So in my case both `Texts` are descendants of the widget I want to tap and in some way they can be considered as siblings each to other. I'll try and I will let you know then.

**Reply by bartekpacia** [reply]
`containing` is a lookahead check – it _looks ahead_ but returns the same `PatrolFinder` that it was called on.

So:

```dart
$(ExampleListTile).containing('User name').containing('user@email.com').containing(FollowButton).tap();
```

will still tap on `ExampleListTile` – it's just that the `ExampleListTile` (that will be tapped on) will meet the conditions you imposed on it with `containing()`.


I think we should mention this in the docs.

**Reply by proximilian** [reply]
Really thanks for the clarifications. It should work in my case then. I will try it and will share tomorrow.

**Reply by proximilian** [reply]
Hello again @bartekpacia,
Sorry for the short delay, but had some other issues. 
Unfortunately, the suggested following way didn't work:  `$(#ExampleListView).$(ExampleWidgetType).containing('text1').containing('text2').tap();`
When trying it, the test stuck without any failure or any log. Just stuck, so I needed to terminated it manually. When I use only one 'containing' method, the test continues without problem, but in that way I cannot tap on the wanted widget that must contain the both `Text` descendants. Any other ideas that come to your mind?

**Reply by OmarYehiaDev** [reply]
I guess you can try `$(#ExampleListView).$(Scrollable). etc...` 
**ListViews** are tricky as it's in the docs:
![image](https://github.com/user-attachments/assets/c2279611-a22f-4b67-9bd4-55ffddcacdcc)
